hooks/shotgun_get_published_file.py

```diff
@@ -0,0 +1,102 @@
+return the first one in the list.
+"""
+import sgtk
+from sgtk import TankError
+
+HookBaseClass = sgtk.get_hook_baseclass()
+
+
+class GetPublishedFile(HookBaseClass):
+    def execute(self, entity_type, entity_id, published_file_entity_type, **kwargs):
+        """
+        Given a Version, PublishedFile or TankPublishedFile,
+        return the PublishedFile or TankPublishedFile linked to it.
+
+        For a Version with multiple PublishedFiles,
+        return the first PublishedFile matching
+        `viewer_extensions` configuration parameter.
+        If no PublishedFile matches the extensions,
+        return the first one in the list.
+
+        :param str entity_type: Version, PublishedFile or TankPublishedFile.
+        :param int entity_id: The id of the entity.
+        :param str published_file_entity_type: PublishedFile or TankPublishedFile.
+        :returns: The selected PublishedFile
+        :raises: TankError if a PublishedFile cannot be found.
+        """
+        if entity_type == "Version":
+            # entity is a Version so try to get the id
+            # of the published file it is linked to
+            if published_file_entity_type == "PublishedFile":
+                v = self.shotgun.find_one("Version", [["id", "is", entity_id]], ["published_files"])
+                if not v.get("published_files"):
+                    raise TankError("Sorry, this can only be used on Versions with an associated PublishedFile.")
+                if len(v["published_files"]) == 1:
+                    publish_id = v["published_files"][0]["id"]
+                    return self.published_file(published_file_entity_type, publish_id)
+                else:  # if there are multiple published files, pick one.
+                    viewer_extensions = self.parent.get_setting("viewer_extensions")
+                    if not viewer_extensions:
+                        raise TankError(
+                            "Sorry, viewer extensions must be provided when a Version has multiple PublishedFiles"
+                        )
+                    published_file_ids = [pf["id"] for pf in v["published_files"]]
+                    published_files = self.sgtk.shotgun.find(
+                        published_file_entity_type,
+                        [["id", "in", published_file_ids]],
+                        ["path", "task", "entity"]
+                    )
+                    for viewer_extension in viewer_extensions:
+                        for published_file in published_files:
+                            path_on_disk = self.parent.get_path_on_disk(published_file)
+                            if path_on_disk and path_on_disk.endswith(".%s" % viewer_extension):
+                                return published_file
+                    return published_files[0]
+
+            else:  # == "TankPublishedFile":
+                v = self.shotgun.find_one("Version", [["id", "is", entity_id]], ["tank_published_file"])
```

Comment:
What is the reason for not doing the extension check here? It seems we could use the same code by having the version field in variable based on the `published_file_entity_type`?

Reply:
The original logic of the code was:
if it's a single published file, pass it on.
If there are multiple, pass the first one.

I changed it to:
If there is a single one, pass it on.
If there are multiple, check viewer_extensions, pass a logical one, or pass the first one.

You want to change it to?
If it's a single one, pass it on only if it corresponds to a viewer extension or raise
If there are multiple, check viewer_extensions, pass the first one that matches, or raise 

Is that right? because for TankPublishedFile, the name of the field, "tank_published_file", indicated to me it was only one file.

Reply:
If `tank_published_file` is a single entity  column, then your code is right. But a comment should be added to explain why we just take whatever is there. If it's a multi-entity, then the same logic should be applied. Makes sense?

Reply:
done.